lib/streamlit/delta_generator.py

```diff
@@ -147,27 +162,49 @@ def __init__(self, container=BlockPath_pb2.BlockPath.MAIN, cursor=None):
         # No relation to `st.container()`.
         self._container = container
 
-        # This is either:
-        # - None: if this is the running DeltaGenerator for a top-level
-        #   container (MAIN or SIDEBAR)
-        # - RunningCursor: if this is the running DeltaGenerator for a
-        #   non-top-level container (created with dg.container())
-        # - LockedCursor: if this is a locked DeltaGenerator returned by some
-        #   other DeltaGenerator method. E.g. the dg returned in dg =
-        #   st.text("foo").
-        #
         # You should never use this! Instead use self._cursor, which is a
         # computed property that fetches the right cursor.
-        #
         self._provided_cursor = cursor
 
+        self.parent = parent
+
+        # Track the DG used for the `with` block.
+        # NOTE: Only the main DG should ever reference this.
+        # You should use the computed property _active_dg instead.
+        self._with_dg = self
+
         # Change the module of all mixin'ed functions to be st.delta_generator,
         # instead of the original module (e.g. st.elements.markdown)
         for mixin in self.__class__.__bases__:
             for (name, func) in mixin.__dict__.items():
                 if callable(func):
                     func.__module__ = self.__module__
 
+    def __enter__(self):
+        # with block started. store the current DG
+        self._main_dg._with_dg = self
+
+    def __exit__(self, type, value, traceback):
+        # with block ended. Reset the current dg back to MAIN
+        # NOTE: This won't handle nested with blocks correctly
+        self._main_dg._with_dg = self._main_dg
+        # Re-raise any exceptions
+        return False
+
+    @property
+    def _active_dg(self):
+        if self == self._main_dg:
+            # `st.button`: Use the current `with` dg
+            return self._with_dg
+        else:
+            # `st.sidebar.button`: Ignore the `with` dg
+            return self
```

Comment:
sorry i was too lazy to look through our meetings and figured i'd just point it out here. are we letting you do containers in sidebar? it looks like in our current container code, you can put a container in a sidebar but with this code, i don't believe you can do `with` for a container in sidebar. or perhaps we need to redefine main_dg?